util/perl/OpenSSL/paramnames.pm

```diff
@@ -0,0 +1,667 @@
+    my @macros = ();
+    my $count = 0;
+    my %reverse;
+
+    # Determine the number for each unique string
+    # Sort the names to improve the change of cache coherency
+    foreach my $name (sort keys %params) {
+        my $val = $params{$name};
+
+        if (substr($val, 0, 1) ne '*' and not defined $reverse{$val}) {
+            $reverse{$val} = $count++;
+        }
+    }
+
+    # Output the defines
+    foreach my $name (keys %params) {
+        my $val = $params{$name};
+        my $def = '#define PIDX_' . $name . ' ';
+
+        if (substr($val, 0, 1) eq '*') {
+            $def .= 'PIDX_' . substr($val, 1);
+        } else {
+            $def .= $reverse{$val};
+        }
+        push(@macros, $def)
+    }
+    return "#define NUM_PIDX $count\n\n" . join("\n", sort @macros);
+}
+
+sub generate_trie {
+    my %trie;
+    my $nodes = 0;
+    my $chars = 0;
+
+    foreach my $name (keys %params) {
+        my $val = $params{$name};
+        if (substr($val, 0, 1) ne '*') {
+            my $cursor = \%trie;
+
+            $chars += length($val);
+            for my $i (0 .. length($val) - 1) {
+                my $c = uc substr($val, $i, 1);
+
+                $c = '_' if $c eq '-' and not $case_sensitive;
+
+                if (not defined $$cursor{$c}) {
+                    $$cursor{$c} = {};
+                    $nodes++;
+                }
+                $cursor = %$cursor{$c};
+            }
+            $$cursor{'val'} = $name;
+        }
+    }
+    #print "\n\n/* $nodes nodes for $chars letters*/\n\n";
+    return %trie;
+}
+
+sub generate_code_from_trie {
+    my $n = shift;
+    my $trieref = shift;
+    my $indent = 1 * $n + 1;
+    my $strcmp = $case_sensitive ? 'strcmp' : 'strcasecmp';
+
+    print "static ossl_inline int ossl_param_find_idx(const char *s)\n{\n" if $n == 0;
+
+    if ($trieref->{'suffix'}) {
+        my $suf = $trieref->{'suffix'};
+
+        printf "%*s if ($strcmp(\"$suf\", s + $n) == 0", $indent, "";
+        if ($case_sensitive) {
+            $suf =~ tr/_/-/;
+            print " || $strcmp(\"$suf\", s + $n) == 0"
+                if ($suf ne $trieref->{'suffix'});
+        }
+        printf ")\n%*s  return PIDX_%s;\n", $indent, "", $trieref->{'name'};
+        printf "%*s break;\n", $indent, "";
+        return;
+    }
+
+    printf "%*sswitch(s\[%d\]) {\n", $indent, "", $n;
+    printf "%*sdefault:\n", $indent, "";
+    for my $l (sort keys %$trieref) {
+        if ($l eq 'val') {
+            printf "%*s break;\n", $indent, "";
+            printf "%*scase '\\0':\n", $indent, "";
+            printf "%*s return PIDX_%s;\n", $indent, "", $trieref->{'val'};
+        } else {
+            printf "%*s break;\n", $indent, "";
+            printf "%*scase '%s':", $indent, "", $l;
+            if (not $case_sensitive) {
+                print "   case '-':" if ($l eq '_');
+                printf "   case '%s':", lc $l if ($l =~ /[A-Z]/);
+            }
+            print "\n";
```

Comment:
Personal nit: I find several cases on the same line harder to read, so:

```suggestion
            printf "%*scase '%s':\n", $indent, "", $l;
            if (not $case_sensitive) {
                print "%*scase '-':\n", $indent, "" if ($l eq '_');
                printf "%*scase '%s':\n", $indent, "", lc $l if ($l =~ /[A-Z]/);
            }
```

Reply:
You aren't meant to read the output C.  Honestly, it is not remotely readable.

Reply:
If we stay case sensitive, this is moot.